### PR TITLE
[tooling] Remember linked declarations

### PR DESCRIPTION
**Pull request by voodoos**
This PR generalizes a mechanism that is used to store "value dependencies" in `cmt` files introduced in https://github.com/ocaml/ocaml/commit/10abdce7b183f5961284ad74ffc27ba5dcaef1d1. It was meant to "track in external tools value declarations between implementations and interfaces".

There are two differences in these changes: 
1. We only store the paired uids and not the entire declaration. The declaration can be retrieved from the corresponding `cmt_uid_to_decl` table (this might require loading a `.cmti` file when the uid is from an interface). 
2. We not only store the paired values, but also modules, modules types, types, etc.

This information will enable external tools to navigate across the declaration graphs of any value, module, type, etc.
Additionally this information can be used to build equivalence classes of declarations. Enabling external tools to perform powerful refactoring by retrieving every usage of every definition whose uids are in the same equivalence class.

This PR compares uids to prevent noisy self-dependencies to be recorded (similarly to what `record_value_dependency` is doing). Since right now uids are [not unique across a compilation unit](https://github.com/ocaml/ocaml/pull/13286) this means that some links between an implementation uid and an interface uid with the same id are missing. Commit 7f5111ba02d5fbbc695ccc021e3933e96c28654a shows the expected test results with https://github.com/ocaml/ocaml/pull/13286 merged.

@alainfrisch I think you were the one to introduce `value_dependencies` which I remove in ec2cc6cd9fbe6eab98d299a4d7465c99051f25d4. Do you now if that feature is actively used today ? I made a search on Sherlocode that return [very few results](https://sherlocode.com/?q=cmt_value_dependencies). One of them is [dead_code_analyser](https://github.com/LexiFi/dead_code_analyzer) which is only available for OCaml < 4.05. The other looks active and compatible with 5.2: [reanalyze](https://github.com/rescript-association/reanalyze). It might be safer to keep both for a while to give time for users to transition ?

cc @lpw25

### CODE REVIEW

File: typing/includemod.ml [hunk 1]
```diff
@@ -897,29 +928,31 @@ and modtype_infos ~core ~in_eq ~loc env ~mark subst id info1 info2 =
...
+        check_modtype_equiv ~core ~direction ~loc env ~mark mty1 mty2 in
   match r with
   | Ok _ as ok -> ok
   | Error e -> Error Error.(Module_type_declaration (diff info1 info2 e))
 
-and check_modtype_equiv ~core ~in_eq ~loc env ~mark mty1 mty2 =
+and check_modtype_equiv ~core ~direction ~loc env ~mark mty1 mty2 =
   let c1 =
-    modtypes ~core ~in_eq:true ~loc env ~mark Subst.identity mty1 mty2
+    let direction = { in_eq = true; positivity = Positive } in
```
Comment (Octachron) [inline typing/includemod.ml]: Thinking more about it, we probably want to play less tricks here and use `at_most_positive direction.positivity` here and `negate` below to keep the correct normalized order of the pair.

File: typing/includemod.ml [hunk 2]
```diff
@@ -427,29 +426,49 @@ type core_relation = {
...
+   [true], we compute a coarser relation [A << B] which is the same as [A <: B]
+   except that module types [T] are checked only for [A.T << B.T] and not the
+   reverse. Thus, we can implement a cheap module type equality check [A.T =
+   B.T] by computing [(A.T << B.T) and (B.T << A.T)], avoiding the exponential
+   slowdown described above.
 *)
+type positivity = Negative | Positive | Strictly_positive
+type directionality = { in_eq : bool; positivity : positivity }
+let strictly_positive = { in_eq = false; positivity = Strictly_positive }
+let negate eq_type =
```
Comment (Octachron) [inline typing/includemod.ml]: Maybe `reverse_direction`  to make the name less context-sensitive?

File: typing/includemod.ml [hunk 3]
```diff
@@ -427,29 +426,49 @@ type core_relation = {
...
+   (B.T <: A.T)], but checking both recursively would lead to an exponential
+   slowdown (see #10598 and #10616). To avoid this issue, when [in_eq] is
+   [true], we compute a coarser relation [A << B] which is the same as [A <: B]
+   except that module types [T] are checked only for [A.T << B.T] and not the
+   reverse. Thus, we can implement a cheap module type equality check [A.T =
+   B.T] by computing [(A.T << B.T) and (B.T << A.T)], avoiding the exponential
+   slowdown described above.
 *)
+type positivity = Negative | Positive | Strictly_positive
+type directionality = { in_eq : bool; positivity : positivity }
```
Comment (Octachron) [inline typing/includemod.ml]: @gasche , would you mind having a look at this part of the PR to check if it is understandable? (since the authorship of this part of the PR has become fuzzy between me and @voodoos )

Comment (gasche) [reply typing/includemod.ml]: I wonder what is the current exchange rate between Merlin PRs and Pattern-Matching Bug PRs.

I'll have a look.

File: typing/includemod.ml (hunk 1 above)
Comment (Octachron) [reply typing/includemod.ml]: In fact, that will still not be sufficient to record pair in the source order for module types in negative position. The current version is probably good enough.

Comment (gasche) [reply typing/includemod.ml]: I don't understand the code here and it smells weird. I would expect one of the following things to be done:
- use `at_most_positive` and `reverse_direction` as you suggest; this restores the intuitive invariant that `negate_mark` and `reverse_direction` (congrats on the naming consistency ;-) are used together
- or, if your last cryptic remark means that you don't trust the positivity information coming from the `direction` argument to `check_modtype_equiv`, then make this clear by only taking an `in_eq : bool` parameter for that construction.

**Review by gasche — COMMENTED** [review]
I had a quick look. My comments:

- I don't know what `Directed | Undirected` in Cmt_format mean, some documentation comment (with an example!) would help.

- Looking at includemod.ml and includemod.mli, the first time we encounter this idea of "positivity" is in the `mark` type, which is scarcely documented, so it's a bit cryptic. I learned more from the implementation comment on `type positivity` in the .ml file, not necessarily because it is very didactic (it focuses on explaining a subtle trick) but because it has meat and something that serves as an example. I wouldn't mind an explanation of this notion of positivity/direction higher in one of the file, with a small example.

- I don't understand why `mark` and `positivity` are two separate mechanisms to trace directions. Maybe they evolve differently for good (unexplained) reasons? Maybe they are partly redundant with each other? Who knows.

- `Strictly_positive` is not documented anywhere that I can see, but my best guess is that it is in fact a combination of `Positive` and a separate `at_toplevel : bool` field being `true`. I suspect that the separate-field presentation would be cleaner.

- (see inline comment about `check_modtype_equiv` which I found supsicious)

File: typing/includemod.ml (hunk 1 above)
Comment (Octachron) [reply typing/includemod.ml]: Switching to `at_most_positive` is probably less surprising indeed.

**Comment by alainfrisch** [comment]
> Do you now if that feature is actively used today ? 

I confirm we still use it as part of our internal version of dead_code_analyzer (not synced with the public version), but don't let this block progress!  I didn't follow developments around `Uid`, but I guess we should be able to switch to using them instead of declarations.

**Comment by Octachron** [comment]
@gasche , the positivity/negativity information is the usual concept of positive/negative position for subtyping  used for parameter variance. In particular, strict positivity is the same in both case sand denotes the absence of double negations.

For modules and Merlin, this information is useful because it corresponds to the only case where the subtyping relationship is matching implementation item with their inferred types to  interface item (compared to the evenly-nested funtor argument cases
```ocaml
module F= functor(_:sig
  module G: functor(_:sig
      val non_strict_pos: int end
    end ) -> sig end
 end) -> struct  let strict_pos = 0 end
 ```
)

We have  added a new commit which better separates the semantics of the marking condition and the positivity tracking. 
Moreover, all information related to "directionality" of the subtyping relationship is now grouped in a single record.

With this refactoring is seemed clearer to simplify the `mark` argument used in various function into a boolean a remove the `mark` type from the interface of `Includemod`. A possible alternative would be to expose the directionality type and its four possible starting values.

File: typing/includemod.ml [hunk 4]
```diff
@@ -134,21 +134,94 @@ module Error = struct
...
+(**
+   When checking inclusion, the [Directionality.t] type tracks the
+   subtyping direction at the syntactic level.
 
-let mark_positive = function
-  | Mark_both | Mark_positive -> true
-  | Mark_negative | Mark_neither -> false
+   The [posivity] field is used in the [cmt_declaration_dependencies] to
+   distinguish between directed and undirected edges, and to avoid recording
+   matched declarations twice.
```
Comment (gasche) [inline typing/includemod.ml:161]: I think you mean [pos] rather than [positivity]. It is unclear to me how the [pos] type "distinguishes between directed and undirected edges". (Or what this means in the first place. I suppose the "edges" are the dependencies, but what does it mean for a dependency to be undirected? Maybe this is better explained somewhere else, but then you could refer to that place.)

Edit: in fact maybe you don't need to document cmt_declaration_dependencies here anymore: now that the `pos` information is necessarily to correctly "mark" the submodules, it is used for strictly more than that.

File: typing/includemod.ml [hunk 5]
```diff
@@ -134,21 +134,94 @@ module Error = struct
...
+
+   The [mark_as_used] field describes if we should record only positive use,
+   any use (because there is no clear implementation side), or none (because we
+   are inside an auxiliary check function.)
+
+   The [in_eq] field is [true] when we are checking both directions inside of
+   module types which allows optimizing module type equality checks. The module
```
Comment (gasche) [inline typing/includemod.ml:168]: I read this sentence as "inside (module types which allows optimizing module type equality checks)", as if certain module types allow this and others don't. I would rephrase this, for example as follows:

> The [in_eq] field is [true] when we are checking one direction of module type equality, rather than subtyping.

File: typing/includemod.ml [hunk 6]
```diff
@@ -134,21 +134,94 @@ module Error = struct
...
+    | Mark_positive ->
+       match d.pos with
+       | Positive | Strictly_positive -> true
+       | Negative -> false
+
+  let normalize_pair dir (elt1, elt2) =
+    match dir.pos with
+    | Negative -> (Cmt_format.Undirected, elt2, elt1)
+    | Positive -> (Cmt_format.Undirected, elt1, elt2)
+    | Strictly_positive -> (Cmt_format.Directed, elt1, elt2)
```
Comment (gasche) [inline typing/includemod.ml]: This function does not strike me as something that clearly should be in this module. (It is not obvious why the notion of direction of module-checking would depend on the notion of dependency edges in the cmt format.) Given that it is only used once that I can see, maybe you could move the logic to the caller, which is still inside includemod.ml but not inside Directionality anymore. Later one maybe we would move the logic further away and change the dependency structures between these modules -- or maybe not.

File: testsuite/tests/uid-deps/link_intf_impl.reference [hunk 7]
```diff
@@ -2,17 +2,16 @@
 Uid dependencies:
 Link_intf_impl.0 <- Link_intf_impl.1
 Link_intf_impl.1 <- Link_intf_impl.0
-Link_intf_impl.4 <- Link_intf_impl.2
+Link_intf_impl.4 <-> Link_intf_impl.2
```
Comment (gasche) [inline testsuite/tests/uid-deps/link_intf_impl.reference]: This commit that refactors mark and directionality together appears to change the testsuite output. This is a bit surprising, is it expected? Is it an improvement?

**Review by gasche — COMMENTED** [review]
I read the new commit that moves "mark" and "positivity" to be in a common submodule, and I like the new result better. Plus you did write some extra documentation, which is nice, thanks! I think that the current state is okay, as far as understandability by others is concerned.

File: testsuite/tests/uid-deps/link_intf_impl.reference (hunk 7 above)
Comment (voodoos) [reply testsuite/tests/uid-deps/link_intf_impl.reference]: It was indeed a regression, fixed in [045312f](https://github.com/ocaml/ocaml/pull/13308/commits/045312f184dd5c4e11d66ad2f022f873c12af9ea)

File: file_formats/cmt_format.mli [hunk 8]
```diff
@@ -50,11 +50,11 @@ and binary_part =
   | Partial_signature_item of signature_item
   | Partial_module_type of module_type
 
+type relation_kind = Directed | Undirected
```
Comment (Octachron) [inline file_formats/cmt_format.mli]: Since this part of the `cmt_format` would be part of the external API, it is probably better to use more expressive name as suggested by @gasche . What about
```ocaml
type dependency_kind = Implementation_to_interface | Any
```
?

**Comment by Octachron** [comment]
With @gasche agreeing with the refactoring inside `Includemod`, the PR looks fine to me. I propose to move forward and merge this PR (and cherry-pick it to 5.3) once we agree on the names for the edge labels in the dependency table.

File: file_formats/cmt_format.mli (hunk 8 above)
Comment (voodoos) [reply file_formats/cmt_format.mli]: Sounds reasonable, I made the change in [01d5a0d](https://github.com/ocaml/ocaml/pull/13308/commits/01d5a0d12dad43d327849ed3296e314fa524f294)

Comment (voodoos) [reply file_formats/cmt_format.mli]: Now that I think about it, `Structure_to_signature`  or `Definition_to_declaration` might be a better choice @Octachron  ? 

These does not only cover cases where the link is between the implementation file and the interface file, but also links between a submodule's structure and its signature.

Comment (Octachron) [reply file_formats/cmt_format.mli]: `Definition_to_declaration` sounds more future proof, and we could expand the `Any` constructor to `Declaration_to_declaration`.

Comment (voodoos) [reply file_formats/cmt_format.mli]: Right, I made these changes.

File: driver/compile_common.ml [hunk 9]
```diff
@@ -58,7 +58,7 @@ let typecheck_intf info ast =
         Format.(fprintf std_formatter) "%a@."
           (Printtyp.printed_signature (Unit_info.source_file info.target))
           sg);
-  ignore (Includemod.signatures info.env ~mark:Mark_both sg sg);
+  ignore (Includemod.signatures info.env ~mark:true sg sg);
```
Comment (Octachron) [inline driver/compile_common.ml:61]: This is the use case for `Includemod.check_implementation`.
